geomesa-filter/src/main/scala/org/locationtech/geomesa/filter/FilterHelper.scala

```diff
@@ -94,283 +164,146 @@ object FilterHelper {
-        val geom = e2.getValue.asInstanceOf[Geometry]
-        val buffer = dwithin.getDistance
-        val bufferedGeom = geom.buffer(buffer)
-        Seq(GeohashUtils.getInternationalDateLineSafeGeometry(bufferedGeom))
-      case bs =>
-        bs.getExpression1.evaluate(null, classOf[Geometry]) match {
-          case g: Geometry => Seq(GeohashUtils.getInternationalDateLineSafeGeometry(g))
-          case _           =>
-            bso.getExpression2.evaluate(null, classOf[Geometry]) match {
-              case g: Geometry => Seq(GeohashUtils.getInternationalDateLineSafeGeometry(g))
-            }
+  /**
+    * Extract geometries from a filter without validating boundaries.
+    *
+    * @param filter filter to evaluate
+    * @param attribute attribute to consider
+    * @return single geometry, if any relevant spatial predicates are present
+    */
+  private def extractGeometry(filter: Filter, attribute: String): Option[Geometry] = filter match {
+    case a: And =>
+      val all = a.getChildren.flatMap(extractGeometry(_, attribute))
+      all.reduceOption[Geometry] { case (g1, g2) => g1.intersection(g2) }
+
+    case o: Or  =>
+      val all = o.getChildren.flatMap(extractGeometry(_, attribute))
+      if (all.length < 2) all.headOption else Some(new GeometryCollection(all.toArray, all.head.getFactory))
+
+    // Note: although not technically required, all known spatial predicates are also binary spatial operators
+    case f: BinarySpatialOperator =>
+      val propertyLiteral = checkOrder(f.getExpression1, f.getExpression2)
+      val withAttribute = propertyLiteral.filter(pl =>  pl.name == null || pl.name == attribute)
+      val geometry = withAttribute.map(_.literal.evaluate(null, classOf[Geometry]))
+      val processedGeometry = geometry.map { geom =>
+        f match {
+          // note: the dwithin should have already between rewritten
+          case dwithin: DWithin => geom.buffer(dwithin.getDistance)
+          case _ => geom
         }
-    }
+      }
+      processedGeometry.map(GeohashUtils.getInternationalDateLineSafeGeometry)
+
+    case _ => None
   }
 
-  // NB: This method assumes that the filters represent a collection of 'and'ed temporal filters.
-  def extractInterval(filters: Seq[Filter], dtField: Option[String], exclusive: Boolean = false): Interval =
-    dtField match {
-      case None      => everywhen
-      case Some(dtf) =>
-        val intervals = filters.map(extractInterval(_, dtf, exclusive))
-        val (s, e) = intervals.fold((minDateTime, maxDateTime))(overlap)
-        if (s > e) null else new Interval(s, e, DateTimeZone.UTC)
+  /**
+    * Extracts intervals from a filter. Intervals will be merged where possible - the resulting sequence
+    * is considered to be a union (i.e. OR)
+    *
+    * @param filter filter to evaluate
+    * @param attribute attribute to consider
+    * @return a sequence of intervals, if any
+    */
+  def extractIntervals(filter: Filter,
```

Comment:
Ok, looks like extractTimes might always return Option[(Long, Long)]?  And this function is just demuxing that?